lib/puma/json.rb

```diff
@@ -0,0 +1,85 @@
+# frozen_string_literal: true
+
+module Puma
+
+  # Puma deliberately avoids the use of the json gem and instead performs JSON
+  # serialization without any external dependencies. In a puma cluster, loading
+  # any gem into the puma master process means that operators cannot use a
+  # phased restart to upgrade their application if the new version of that
+  # application uses a different version of that gem. The json gem in
+  # particular is additionally problematic because it leverages native
+  # extensions. If the puma master process relies on a gem with native
+  # extensions and operators remove gems from disk related to old releases,
+  # subsequent phased restarts can fail.
+  #
+  # The implementation of JSON serialization in this module is not designed to
+  # be particularly full-featured or fast. It just has to handle the few places
+  # where Puma relies on JSON serialization internally.
+
+  module JSON
+    class SerializationError < StandardError; end
+
+    class << self
+      def generate(value)
+        case value
+        when Array, Hash
+          parts = []
+          serialize_value parts, value
+          parts.join ''
+        else
+          raise SerializationError, "Could not serialize object of type #{value.class}"
+        end
+      end
+
+      private
+
+      def serialize_value(output, value)
+        case value
+        when true
+          output << 'true'
+        when false
+          output << 'false'
+        when Array
+          output << '['
+          value.each_with_index do |member, index|
+            output << ',' if index != 0
+            serialize_value output, member
+          end
+          output << ']'
+        when Hash
+          output << '{'
+          value.each_with_index do |(k, v), index|
+            output << ',' if index != 0
+            serialize_object_key output, k
+            output << ':'
+            serialize_value output, v
+          end
+          output << '}'
+        when String
+          serialize_string output, value
+        when Integer, Float
+          output << value.to_s
+        else
+          raise SerializationError, "Unexpected value of type #{value.class}"
+        end
+      end
+
+      def serialize_string(output, value)
+        output << '"'
+        output << value.gsub(/[\\"]/, '\\' => '\\\\', '"' => '\\"')
+        output << '"'
+      end
+
+      def serialize_object_key(output, value)
+        case value
+        when Symbol
+          serialize_string output, value.to_s
+        when String
+          serialize_string output, value
```

Comment:
In the past I used inspect for strings and symbols:
```ruby
when Symbol, String
  output << value.to_s.inspect
```
or 
```ruby
when Symbol
  output << value.to_s.inspect
when String
  output << value.inspect
```
I think it does all the needed processing?

Reply:
I'll dig deeper into this. I realized I'm not handling all of the different escape sequences correctly. Newline characters, for example, are written literally instead of `\n`.

I'll have to double check that everything that `.inspect` emits is valid JSON. If it is, I'll use it, otherwise, I'll handle the special characters in a different way.

Reply:
Thanks.  Your comment said
```
# The implementation of JSON serialization in this module is not designed to
# be particularly full-featured or fast. It just has to handle the few places
# where Puma relies on JSON serialization internally.
```
I suspect you know better than any of us what Puma sends to `JSON.generate` (having just reviewed it), so I'll trust your judgement as to what edge cases it should handle.

Reply:
Added 98b3408b4de75652aca8cfc66b08a6deee6da794 to handle control characters correctly and added 28e9d8ffe5ce7e51bced113e8c266182e58f01aa to simplify the handling of `String` and `Symbol`.

AFAICT, there's no guarantee that `.inspect` produces valid JSON (or will always produce valid JSON into the future), and adding support for control characters explicitly ended up not being so bad. I think the cost of adding the cases for the different characters is small, and it makes the implementation more closely match the language of the specification. I prefer that to having to worry about some edge case down the road where `.inspect` produces something that JSON parsers don't like.